### PR TITLE
add finals

### PR DESCRIPTION
**Pull request by marvin-bitterlich**
many places have final problems.

This saves performance and we can always remove the final again later

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 1]
```diff
@@ -211,7 +211,7 @@ public AxisAlignedBB getBoundingBox(final IBlockState state, final IBlockAccess
     @SideOnly(Side.CLIENT)
     @Override
     @Deprecated
-    public boolean shouldSideBeRendered(IBlockState blockState, IBlockAccess blockAccess, BlockPos pos, EnumFacing side)
+    public boolean shouldSideBeRendered(final IBlockState blockState, final IBlockAccess blockAccess, final BlockPos pos, final EnumFacing side)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:214]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/colony/Colony.java [hunk 2]
```diff
@@ -559,7 +559,7 @@ public void incrementStatistic(@NotNull String statistic)
      * @param statistic the statistic.
      * @return amount of statistic.
      */
-    private int getStatisticAmount(@NotNull String statistic)
+    private int getStatisticAmount(@NotNull final String statistic)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/Colony.java:562]: MAJOR Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowSkeleton.java [hunk 3]
```diff
@@ -35,7 +35,7 @@ public AbstractWindowSkeleton(final String resource)
      * @param id     Button ID.
      * @param action Consumer with the action to be performed.
      */
-    public final void registerButton(String id, Runnable action)
+    public final void registerButton(final String id, final Runnable action)
     {
         registerButton(id, (button) -> action.run());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowSkeleton.java:40]: MINOR Remove the parentheses around the "button" parameter (sonar.java.source not set. Assuming 8 or greater.) [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1611)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java [hunk 4]
```diff
@@ -146,7 +146,7 @@ public AbstractBuilding getTask()
      * @param addToList if is in warehouse should add to the list?
      * @return true if has something in warehouse to deliver.
      */
-    public boolean checkInWareHouse(@NotNull final AbstractBuilding buildingEntry, boolean addToList)
+    public boolean checkInWareHouse(@NotNull final AbstractBuilding buildingEntry, final boolean addToList)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:149]: MAJOR Refactor this method to reduce its Cognitive Complexity from 25 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/client/gui/GuiFactory.java [hunk 5]
```diff
@@ -35,7 +35,7 @@ public void initialize(Minecraft minecraftInstance)
      */
     @Override
     @Deprecated
-    public RuntimeOptionGuiHandler getHandlerFor(RuntimeOptionCategoryElement element)
+    public RuntimeOptionGuiHandler getHandlerFor(final RuntimeOptionCategoryElement element)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/GuiFactory.java:38]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java [hunk 6]
```diff
@@ -54,7 +54,7 @@ public static void placeConstructionTape(@NotNull AbstractBuilding building, @No
      * @param world     the world.
      */
 
-    public static void placeConstructionTape(@NotNull WorkOrderBuildDecoration workOrder, @NotNull World world)
+    public static void placeConstructionTape(@NotNull final WorkOrderBuildDecoration workOrder, @NotNull final World world)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java:57]: MAJOR Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/commands/AbstractSingleCommand.java [hunk 7]
```diff
@@ -115,7 +115,7 @@ public boolean isPlayerOpped(@NotNull final ICommandSender sender, String cmdNam
      * @param theCommand which command to check if the player can use it
      * @return boolean
      */
-    public boolean canCommandSenderUseCommand(Commands theCommand)
+    public boolean canCommandSenderUseCommand(final Commands theCommand)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/commands/AbstractSingleCommand.java:118]: MAJOR Reduce the number of returns of this method 13, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 8]
```diff
@@ -256,7 +256,7 @@ private void requestBlockToBuildingIfRequired(BuildingBuilder building, IBlockSt
     /**
      * Adds entities to the builder building if he needs it.
      */
-    private void requestEntityToBuildingIfRequired(Template.EntityInfo entityInfo)
+    private void requestEntityToBuildingIfRequired(final Template.EntityInfo entityInfo)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java:259]: MAJOR Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java [hunk 9]
```diff
@@ -268,7 +268,7 @@ public AxisAlignedBB getBoundingBox(final IBlockState state, final IBlockAccess
     @SideOnly(Side.CLIENT)
     @Override
     @Deprecated
-    public boolean shouldSideBeRendered(IBlockState blockState, IBlockAccess blockAccess, BlockPos pos, EnumFacing side)
+    public boolean shouldSideBeRendered(final IBlockState blockState, final IBlockAccess blockAccess, final BlockPos pos, final EnumFacing side)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java:271]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 10]
```diff
@@ -713,7 +713,7 @@ else if (targetFile.exists())
         {
             Files.copy(file.toPath(), targetFile.toPath());
         }
-        catch (IOException e)
+        catch (final IOException e)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:716]: MAJOR Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 11]
```diff
@@ -713,7 +713,7 @@ else if (targetFile.exists())
...
         {
             e.printStackTrace();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:718]: MINOR Use a logger to log this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/minecolonies/coremod/commands/GetColonyAndCitizen.java [hunk 12]
```diff
@@ -91,18 +91,18 @@ else if (args.length >= TOO_MANY_ARGUMENTS_LENGTH)
      * @param args     The given arguments
      * @return Returns citizen ID
      */
-    public static int getCitizenId(@NotNull int colonyId, @NotNull final String... args)
+    public static int getCitizenId(@NotNull final int colonyId, @NotNull final String... args)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/commands/GetColonyAndCitizen.java:94]: MAJOR The Cyclomatic Complexity of this method "getCitizenId" is 16 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Refactor this method to reduce its Cognitive Complexity from 25 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java [hunk 13]
```diff
@@ -913,7 +913,7 @@ public NBTTagList writeToNBT(NBTTagList nbtTagListIn)
         {
             if (!(this.offHandInventory.get(k)).isEmpty())
             {
-                NBTTagCompound nbttagcompound2 = new NBTTagCompound();
+                final NBTTagCompound nbttagcompound2 = new NBTTagCompound();
                 nbttagcompound2.setByte("Slot", (byte) (k + 150));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java:917]: MAJOR Assign this magic number 150 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java [hunk 14]
```diff
@@ -1034,13 +1034,13 @@ public void setItemStack(ItemStack itemStackIn)
...
     {
         label19:
 
-        for (List<ItemStack> list : this.allInventories)
+        for (final List<ItemStack> list : this.allInventories)
         {
-            Iterator<ItemStack> iterator = list.iterator();
+            final Iterator<ItemStack> iterator = list.iterator();
 
             while (true)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java:1045]: MAJOR Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)

File: src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java [hunk 15]
```diff
@@ -279,19 +279,19 @@ public boolean hasCustomName()
      * @param itemNBT     The NBT data to match, null ignores.
      * @return The number of items removed from the inventory.
      */
-    public int clearMatchingItems(@javax.annotation.Nullable Item itemIn, int metadataIn, int removeCount, @javax.annotation.Nullable NBTTagCompound itemNBT)
+    public int clearMatchingItems(@javax.annotation.Nullable final Item itemIn, final int metadataIn, final int removeCount, @javax.annotation.Nullable final NBTTagCompound itemNBT)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java:282]: MAJOR The Cyclomatic Complexity of this method "clearMatchingItems" is 26 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Refactor this method to reduce its Cognitive Complexity from 45 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
MAJOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

File: src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java [hunk 16]
```diff
@@ -902,7 +902,7 @@ public NBTTagList writeToNBT(NBTTagList nbtTagListIn)
         {
             if (!(this.armorInventory.get(j)).isEmpty())
             {
-                NBTTagCompound nbttagcompound1 = new NBTTagCompound();
+                final NBTTagCompound nbttagcompound1 = new NBTTagCompound();
                 nbttagcompound1.setByte("Slot", (byte) (j + 100));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java:906]: MAJOR Assign this magic number 100 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java [hunk 17]
```diff
@@ -1034,13 +1034,13 @@ public void setItemStack(ItemStack itemStackIn)
      * @param itemStackIn the stack to be searched for.
      * @return true if it exists.
      */
-    public boolean hasItemStack(ItemStack itemStackIn)
+    public boolean hasItemStack(final ItemStack itemStackIn)
     {
         label19:
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java:1039]: MAJOR Refactor the code to remove this label and the need for it. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALabelsShouldNotBeUsedCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 18]
```diff
@@ -203,7 +203,7 @@ protected void onException(final RuntimeException e)
             // fix for printing the actual exception
             e.printStackTrace();
         }
-        catch (RuntimeException exp)
+        catch (final RuntimeException exp)
         {
             Log.getLogger().error("Welp reporting crashed:");
             exp.printStackTrace();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:209]: MINOR Use a logger to log this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 19]
```diff
@@ -203,7 +203,7 @@ protected void onException(final RuntimeException e)
             // fix for printing the actual exception
             e.printStackTrace();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:204]: MINOR Use a logger to log this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 20]
```diff
@@ -203,7 +203,7 @@ protected void onException(final RuntimeException e)
...
         }
-        catch (RuntimeException exp)
+        catch (final RuntimeException exp)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:206]: MAJOR Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 202 issues
* ![MAJOR][MAJOR] 125 major
* ![MINOR][MINOR] 51 minor
* ![INFO][INFO] 26 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [WindowHireWorker.java#L125](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java#L125): Reduce this anonymous class number of lines from 31 to at most 20, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [CitizenData.java#L23](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L23): class "CitizenData" has 43 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.226 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L48](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L48): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 44 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L48](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L48): class "Colony" has 83 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L182](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L182): Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1026](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1026): Assign this magic number 20 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1027](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1027): Assign this magic number 60 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1151](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1151): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1153](https://github.com/Minecolonies/minecolonies/blob/34830fbee534f9cb8da5b7df9d8b1179d26b00b2/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1153): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by xavierh** [comment]
nice
LGTM

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1083/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1083/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**